File: ruby/loadbalancer/TCPLoadBalancer.rb
require 'socket'
require 'timeout'
require 'yaml'
require 'async'
require_relative 'roundRobinBalancer'

class TCPLoadBalancer
  def initialize(config)
    @listen_host = config['listen_host']
    @listen_port = config['listen_port']
    @load_balancer = RoundRobinBalancer.new(
      config['backend_servers'],
      timeout: config['timeout']
    )
    @health_check_interval = config['health_check_interval']
    @shutdown = false
    @health_check_thread = nil
    @server = nil
  end

  def start
    Async do |task|
      @health_check_task = task.async { run_health_checks }

      @server = TCPServer.new(@listen_host, @listen_port)
      puts "Load balancer is listening on port #{@listen_port}..."

      task.async do
        loop do
          break if @shutdown

          begin
            client = @server.accept
            task.async { handle_connection_with_failover(client) }
          rescue IOError, Errno::EBADF => e
            break if @shutdown
            puts "Error accepting client connection: #{e.message}"
          end
        end
      end
    end
  ensure
    stop
  end

  def stop
    @shutdown = true
    @health_check_thread&.kill
    @server&.close rescue nil
    puts "Load balancer has stopped."
  end

  private

  def run_health_checks
    loop do
      break if @shutdown
      sleep @health_check_interval
      puts "Running health checks..."
      @load_balancer.health_check
    end
  end

  def handle_connection_with_failover(client)
    attempt_count = 0
    backend_server = nil

    begin
      attempt_count += 1
      server = @load_balancer.next_server
      puts "Redirecting to #{server.address} (Attempt #{attempt_count})"

      backend_server = TCPSocket.new(server.ip, server.port)
      relay_data(client, backend_server)
    rescue => e
      puts "Error: #{e.message}"
      if attempt_count < @load_balancer.instance_variable_get(:@servers).size
        puts "Attempting failover to the next server..."
        retry
      else
        puts "All servers failed. Returning 503 to client."
        client.puts "503 Service Unavailable"
      end
    ensure
      client.close if client && !client.closed?
      backend_server&.close if backend_server && !backend_server.closed?
    end
  end

  def relay_data(client, backend_server)
    sockets = [client, backend_server]
    loop do
      break if sockets.any?(&:closed?)

      ready_sockets = IO.select(sockets)
      break if ready_sockets.nil?

      ready_sockets[0].each do |socket|
        begin
          data = socket.read_nonblock(4096)
          target_socket = (socket == client) ? backend_server : client
          if target_socket && !target_socket.closed?
            target_socket.write_nonblock(data)
          end
        rescue IO::WaitReadable, IO::WaitWritable
          next
        rescue EOFError, IOError, Errno::ECONNRESET, Errno::EPIPE
          socket.close unless socket.closed?
        end
      end
    end
  end
end

File: go/loadbalancer/tcp_load_balancer.go
package main

import (
	"context"
	"io"
	"log"
	"net"
	"sync"
	"time"
)

type TCPLoadBalancer struct {
	listenHost           string
	listenPort           string
	loadBalancer         *RoundRobinBalancer
	healthCheckInterval  time.Duration
	shutdown             bool
	healthCheckWaitGroup sync.WaitGroup
	server               net.Listener
}

func NewTCPLoadBalancer(config map[string]interface{}) *TCPLoadBalancer {
	servers := config["backend_servers"].([]string)
	timeout := time.Duration(config["timeout"].(int)) * time.Second
	healthCheckInterval := time.Duration(config["health_check_interval"].(int)) * time.Second

	return &TCPLoadBalancer{
		listenHost:          config["listen_host"].(string),
		listenPort:          config["listen_port"].(string),
		loadBalancer:        NewRoundRobinBalancer(servers, timeout),
		healthCheckInterval: healthCheckInterval,
	}
}

func (lb *TCPLoadBalancer) Start() {
	var err error
	lb.server, err = net.Listen("tcp", lb.listenHost+":"+lb.listenPort)
	if err != nil {
		log.Println("Error starting server:", err)
		return
	}
	defer lb.server.Close()

	log.Printf("Load balancer is listening on port %s...\n", lb.listenPort)

	lb.healthCheckWaitGroup.Add(1)
	go lb.runHealthChecks()

	for {
		if lb.shutdown {
			break
		}

		conn, err := lb.server.Accept()
		if err != nil {
			if lb.shutdown {
				break
			}
			log.Println("Error accepting client connection:", err)
			continue
		}
		go lb.handleConnectionWithFailover(conn)
	}
}

func (lb *TCPLoadBalancer) Stop() {
	lb.shutdown = true
	lb.server.Close()
	lb.healthCheckWaitGroup.Wait()
	log.Println("Load balancer has stopped.")
}

func (lb *TCPLoadBalancer) runHealthChecks() {
	defer lb.healthCheckWaitGroup.Done()
	ticker := time.NewTicker(lb.healthCheckInterval)
	defer ticker.Stop()

	for range ticker.C {
		if lb.shutdown {
			return
		}
		log.Println("Running health checks...")
		lb.loadBalancer.HealthCheck()
	}
}

func (lb *TCPLoadBalancer) handleConnectionWithFailover(client net.Conn) {
	defer client.Close()

	attemptCount := 0
	var backendServer net.Conn

	for {
		attemptCount++
		server, err := lb.loadBalancer.NextServer()
		if err != nil {
			log.Println("All servers failed. Returning 503 to client.")
			client.Write([]byte("503 Service Unavailable\n"))
			return
		}

		if attemptCount > lb.loadBalancer.HealthyServerCount() {
			log.Println("Exceeded maximum retry attempts. Returning 503 to client.")
			client.Write([]byte("503 Service Unavailable\n"))
			return
		}

		log.Printf("Redirecting to %s (Attempt %d)\n", server.Address(), attemptCount)
		backendServer, err = net.Dial("tcp", server.Address())
		if err == nil {
			break
		}
		log.Println("Error:", err)
	}

	defer backendServer.Close()
	lb.relayData(client, backendServer)
}

func (lb *TCPLoadBalancer) relayData(client, backendServer net.Conn) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	var wg sync.WaitGroup
	wg.Add(2)

	go func() {
		defer wg.Done()
		io.Copy(backendServer, client)
		cancel()
	}()

	go func() {
		defer wg.Done()
		io.Copy(client, backendServer)
		cancel()
	}()

	waitForCompletion(ctx, &wg)
}

func waitForCompletion(ctx context.Context, wg *sync.WaitGroup) {
	select {
	case <-ctx.Done():
	case <-func() chan struct{} {
		done := make(chan struct{})
		go func() {
			wg.Wait()
			close(done)
		}()
		return done
	}():
	}
}

func main() {
	config := map[string]interface{}{
		"listen_host":           "127.0.0.1",
		"listen_port":           "8000",
		"backend_servers":       []string{"127.0.0.1:8001", "127.0.0.1:8002"},
		"timeout":               5,
		"health_check_interval": 10,
	}

	lb := NewTCPLoadBalancer(config)
	lb.Start()
}
